build.rs

```diff
@@ -18,21 +22,77 @@ fn main() {
         let lapacke = feature!("LAPACKE");
         let source = PathBuf::from("source");
         let output = PathBuf::from(variable!("OUT_DIR").replace(r"\", "/"));
+        // The target triple provided by cargo --target is not compatible with
+        // OpenBLAS targets.
         env::remove_var("TARGET");
-        run(
-            Command::new("make")
-                .args(&["libs", "netlib", "shared"])
-                .arg(format!("BINARY={}", binary!()))
-                .arg(format!("{}_CBLAS=1", switch!(cblas)))
-                .arg(format!("{}_LAPACKE=1", switch!(lapacke)))
-                .arg(format!("-j{}", variable!("NUM_JOBS")))
-                .current_dir(&source),
-        );
+        // The working directory for both 'make' and 'make install'
+        let make_working_dir: PathBuf;
+        let mut make_build_cmd = Command::new("make");
+        make_build_cmd
+            .args(&["libs", "netlib", "shared"])
+            .arg(format!("BINARY={}", binary!()))
+            .arg(format!("{}_CBLAS=1", switch!(cblas)))
+            .arg(format!("{}_LAPACKE=1", switch!(lapacke)))
+            .arg(format!("-j{}", variable!("NUM_JOBS")));
+        // Options for cross compilation
+        match env::var("OPENBLAS_TARGET") {
+            Ok(target) => {
+                make_build_cmd.arg(format!("TARGET={}", target));
+
+                // Since target is user-specified, we relax strictness on case.
+                let canonical_target = target.to_uppercase();
+                let canonical_target_path = Path::new(&canonical_target);
+
+                // Check if there's already a dedicated folder for compiling this target.
+                // If not, create it by duplicating the source folder.
+                if !Path::new(&canonical_target).exists() {
+                    let canonical_target_tmp = format!("{}_TMP", canonical_target);
+                    // If the tmp folder exists, the copy must have failed during
+                    // a previous invocation so remove it.
+                    let canonical_target_tmp_path = Path::new(&canonical_target_tmp);
+                    if canonical_target_tmp_path.exists() {
+                        fs::remove_dir_all(canonical_target_tmp_path).unwrap();
+                    }
+                    run(
+                        Command::new("cp")
+                            .arg("-R")
+                            .arg("source")
+                            .arg(&canonical_target_tmp)
+                    );
+                    // Remove any existing compiled files since the source folder
+                    // is still used for non-cross-compilations.
+                    run(
```

Comment:
Should not be needed if we switch to the aforementioned treat-all-targets-identically approach.